diff --git a/media/libmediaplayerservice/StagefrightRecorder.cpp b/media/libmediaplayerservice/StagefrightRecorder.cpp
index 2aba3261e7..2052a6e6a0 100644
--- a/media/libmediaplayerservice/StagefrightRecorder.cpp
+++ b/media/libmediaplayerservice/StagefrightRecorder.cpp
@@ -1127,6 +1127,9 @@ status_t StagefrightRecorder::setParameter(
         }
     } else if (key == "log-session-id") {
         return setLogSessionId(value);
+    } else if (key == "OplusUserData") {
+        mOplusUserData = value.c_str();
+        return OK;
     } else {
         ALOGE("setParameter: failed to find key %s", key.c_str());
     }
@@ -2166,7 +2169,7 @@ status_t StagefrightRecorder::setupVideoEncoder(
     sp<MediaCodecSource> encoder = MediaCodecSource::Create(
             mLooper, format, cameraSource, mPersistentSurface, flags);
     if (encoder == NULL) {
-        ALOGE("Failed to create video encoder");
+        ALOGE("[PATCH OPLUS] FATAL: Failed to create video encoder!");
         // When the encoder fails to be created, we need
         // release the camera source due to the camera's lock
         // and unlock mechanism.
@@ -2314,6 +2317,11 @@ void StagefrightRecorder::setupMPEG4orWEBMMetaData(sp<MetaData> *meta) {
         (*meta)->setInt32(kKeyEmptyTrackMalFormed, true);
         (*meta)->setInt32(kKey4BitTrackIds, true);
     }
+
+    // [PATCH OPLUSHDR] Pass metadata to MPEG4Writer
+    if (!mOplusUserData.empty()) {
+        (*meta)->setCString(kKeyOplusUserData, mOplusUserData.c_str());
+    }
 }
 
 status_t StagefrightRecorder::pause() {
diff --git a/media/libstagefright/MPEG4Writer.cpp b/media/libstagefright/MPEG4Writer.cpp
index 2dac9e95a3..109ffe66f9 100644
--- a/media/libstagefright/MPEG4Writer.cpp
+++ b/media/libstagefright/MPEG4Writer.cpp
@@ -1542,7 +1542,11 @@ void MPEG4Writer::writeMvhdBox(int64_t durationUs) {
 void MPEG4Writer::writeMoovBox(int64_t durationUs) {
     beginBox("moov");
     writeMvhdBox(durationUs);
-    if (mAreGeoTagsAvailable) {
+
+    const char *oplusData = nullptr;
+    bool hasOplusData = (mStartMeta != nullptr && mStartMeta->findCString(kKeyOplusUserData, &oplusData) && oplusData != nullptr);
+
+    if (mAreGeoTagsAvailable || hasOplusData) {
         writeUdtaBox();
     }
     writeMoovLevelMetaBox();
@@ -5498,7 +5502,18 @@ void MPEG4Writer::Track::writeCo64Box() {
 
 void MPEG4Writer::writeUdtaBox() {
     beginBox("udta");
-    writeGeoDataBox();
+
+    if (mAreGeoTagsAvailable) {
+        writeGeoDataBox();
+    }
+
+    const char *oplusData = nullptr;
+    if (mStartMeta != nullptr && mStartMeta->findCString(kKeyOplusUserData, &oplusData) && oplusData != nullptr) {
+        beginBox("oplu");
+        writeCString(oplusData);
+        endBox();
+    }
+
     endBox();
 }
 
diff --git a/media/libstagefright/include/media/stagefright/MetaDataBase.h b/media/libstagefright/include/media/stagefright/MetaDataBase.h
index 0a5d4019fd..687f611fa6 100644
--- a/media/libstagefright/include/media/stagefright/MetaDataBase.h
+++ b/media/libstagefright/include/media/stagefright/MetaDataBase.h
@@ -68,6 +68,7 @@ enum {
     kKeyThumbnailAV1C     = 'tav1',  // raw data
     kKeyD263              = 'd263',  // raw data
     kKeyOpusHeader        = 'ohdr',  // raw data
+    kKeyOplusUserData     = 'opud',  // cstring
     kKeyOpusCodecDelay    = 'ocod',  // uint64_t (codec delay in ns)
     kKeyOpusSeekPreRoll   = 'ospr',  // uint64_t (seek preroll in ns)
     kKeyVp9CodecPrivate   = 'vp9p',  // raw data (vp9 csd information)
diff --git a/services/camera/libcameraservice/Android.bp b/services/camera/libcameraservice/Android.bp
index f9cbc0e6d8..2bc41779f1 100644
--- a/services/camera/libcameraservice/Android.bp
+++ b/services/camera/libcameraservice/Android.bp
@@ -241,6 +241,7 @@ cc_library {
         "-Wextra",
         "-Werror",
         "-Wno-ignored-qualifiers",
+        "-DALLOW_NONINCREASING_TIMESTAMPS",
     ],
 }
 
diff --git a/services/camera/libcameraservice/device3/Camera3Stream.cpp b/services/camera/libcameraservice/device3/Camera3Stream.cpp
index 6a44ee38be..aeb60f0105 100644
--- a/services/camera/libcameraservice/device3/Camera3Stream.cpp
+++ b/services/camera/libcameraservice/device3/Camera3Stream.cpp
@@ -778,12 +778,16 @@ status_t Camera3Stream::returnBuffer(const camera_stream_buffer &buffer,
 
     // Buffer status may be changed, so make a copy of the stream_buffer struct.
     camera_stream_buffer b = buffer;
+#ifdef ALLOW_NONINCREASING_TIMESTAMPS
+    (void)timestampIncreasing;
+#else
     if (timestampIncreasing && timestamp != 0 && timestamp <= mLastTimestamp
             && b.status != CAMERA_BUFFER_STATUS_ERROR) {
         ALOGE("%s: Stream %d: timestamp %" PRId64 " is not increasing. Prev timestamp %" PRId64,
                 __FUNCTION__, mId, timestamp, mLastTimestamp);
         b.status = CAMERA_BUFFER_STATUS_ERROR;
     }
+#endif
     mLastTimestamp = timestamp;
 
     /**
